gk/main.c

```diff
@@ -918,20 +944,26 @@ process_pkts_front(uint8_t port_front, uint8_t port_back,
 				 * Notice that one needs to update
 				 * the Ethernet header.
 				 *
-				 * TODO Add sequential lock to deal with the
+				 * Use sequential lock to deal with the
 				 * concurrency between GK and LLS on the cached
 				 * Ethernet header.
 				 */
+
+				bool stale;
+
 				eth_cache = fib->u.gateway.eth_cache;
-				if (eth_cache == NULL || eth_cache->stale) {
+				if (eth_cache == NULL) {
+					drop_packet(pkt);
+					continue;
+				}
+
+				stale = pkt_copy_cached_eth_header(
+					pkt, eth_cache);
+				if (stale) {
```

Comment:
Once you combine the lines above as a prior comment explains, you'll be able to combine this whole `if` statement with the previous `if` statement. A similar thing may happen in other places.